core/arch/arm/kernel/secure_partition.c

```diff
@@ -407,6 +407,29 @@ static TEE_Result check_fdt(const void * const fdt, const TEE_UUID *uuid)
 	return TEE_SUCCESS;
 }
 
+static size_t calc_fdt_extra_size(const void * const fdt)
+{
+	int node = 0;
+	int subnode = 0;
+	size_t extra_size = 0;
+
+	node = fdt_node_offset_by_compatible(fdt, 0,
+					     "arm,ffa-manifest-memory-regions");
+
+	/*
+	 * The base-address field is used to pass the VA of the mapped region to
+	 * the SP. Normally this would be just an in-place write, but if the
+	 * field is missing, we have to allocate extra space for it (when
+	 * copying the fdt) so we can fill it in later.
+	 */
+	fdt_for_each_subnode(subnode, fdt, node) {
+		if (!fdt_get_property(fdt, subnode, "base-address", NULL))
+			extra_size += sizeof("base-address") + sizeof(fdt64_t);
```

Comment:
I just found a test case when this is wrong. This solution makes some assumptions about the fdt internal representation which might not always be true. I'm not sure if there is any way to ask libfdt how much space we should reserve for a property we want to add later.

Realistically this fine grained byte counting is not necessary. The allocated space for the `struct sp_ffa_init_info` plus the fdt will be one page anyway, so it might be better to simply use a large enough fixed value at `fdt_open_into()`. Also, this memory will be freed right after the SP is initialised, so we're not wasting space. I'll refactor this part.

Reply:
Refactored.

Reply:
@jenswi-linaro could you take a look at these modifications? Thanks!